src/nunavut/lang/cpp/templates/_composite_type.j2

```diff
@@ -31,8 +31,69 @@
 {% if composite_type.deprecated -%}
 [[deprecated("{{ composite_type }} is reaching the end of its life; there may be a newer version available")]]
 {% endif -%}
-struct {{composite_type|short_reference_name}} final
+template <class T, class Allocator = std::allocator<T>>
+struct {{composite_type|short_reference_name}}_Impl final
 {
+    // Default constructor
+    {{composite_type|short_reference_name}}_Impl(const Allocator& allocator = Allocator()){%if composite_type.fields_except_padding %} :{%endif%}
+    {%- for field in composite_type.fields_except_padding %}
+        {{ field | id }}{{ field | value_initializer(SpecialMethods.DefaultConstructorWithOptionalAllocator) }}{%if not loop.last %},{%endif %}
+    {%- endfor %}
+    { }
+
+    // Destructor
+    ~{{composite_type|short_reference_name}}_Impl() = default;
+
+    // Copy constructor
+    {{composite_type|short_reference_name}}_Impl(const {{composite_type|short_reference_name}}_Impl& rhs){%if composite_type.fields_except_padding %} :{%endif%}
+    {%- for field in composite_type.fields_except_padding %}
+        {{ field | id }}{{ field | value_initializer(SpecialMethods.CopyConstructor) }}{%if not loop.last %},{%endif %}
+    {%- endfor %}
+    {
+    }
+
+    // Copy constructor with allocator
+    {{composite_type|short_reference_name}}_Impl(const {{composite_type|short_reference_name}}_Impl& rhs, const Allocator& allocator){%if composite_type.fields_except_padding %} :{%endif%}
+    {%- for field in composite_type.fields_except_padding %}
+        {{ field | id }}{{ field | value_initializer(SpecialMethods.CopyConstructorWithAllocator) }}{%if not loop.last %},{%endif %}
+    {%- endfor %}
+    {
+    }
+
+    // Move constructor
+    {{composite_type|short_reference_name}}_Impl({{composite_type|short_reference_name}}_Impl&& rhs){%if composite_type.fields_except_padding %} :{%endif%}
+    {%- for field in composite_type.fields_except_padding %}
+        {{ field | id }}{{ field | value_initializer(SpecialMethods.MoveConstructor) }}{%if not loop.last %},{%endif %}
+    {%- endfor %}
+    {
+    }
+
+    // Move constructor with allocator
+    {{composite_type|short_reference_name}}_Impl({{composite_type|short_reference_name}}_Impl&& rhs, const Allocator& allocator){%if composite_type.fields_except_padding %} :{%endif%}
+    {%- for field in composite_type.fields_except_padding %}
+        {{ field | id }}{{ field | value_initializer(SpecialMethods.MoveConstructorWithAllocator) }}{%if not loop.last %},{%endif %}
+    {%- endfor %}
+    {
+    }
+
+    // Copy assignment
+    {{composite_type|short_reference_name}}_Impl& operator=(const {{composite_type|short_reference_name}}_Impl& rhs)
+    {
+    {%- for field in composite_type.fields_except_padding %}
+        {{ field | id }} = rhs.{{ field | id }};
+    {%- endfor %}
+        return *this;
+    }
+
+    // Move assignment
```

Comment:
For the special methods that don't take an allocator I suppose "= default" would supply the same implementation, right?  If that's correct I could shorten the code-gen for those cases.